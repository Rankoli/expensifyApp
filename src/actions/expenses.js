import uuid from 'uuid';
import database from '../firebase/firebase';
import axios from 'axios';
import Api from '../../server/api';

// ADD_EXPENSE
 export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
  });

  export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
        description = '', note = '', amount = 0, createdAt = 0
      } = expenseData
      const expense = {description, note, amount, createdAt};

    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
    };
  };

  export const testReducer = () => ({
    type: 'TEST_T'
  });

  export const tryingToConnectDatabase = () => {
    return (dispatch) => {
        return Api.post("/Login",{email: 'rankoli3@gmail.com',pass: '123456'}).then((Response) => {
          dispatch(testReducer());
          console.log(Response.data.d);
          const user = JSON.parse(Response.data.d);
          console.log(user.Email, user.Password);
        }).catch((error) => {
          console.log(error);
        })
    };
  };

  // REMOVE_EXPENSE
 export  const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
      id: id
  });

  export const startRemoveExpense = ({ id = {}}) => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
        dispatch(removeExpense({id}));
      });
    };
  };

  // EDIT_EXPENSE
 export  const editExpense = (id,updates) => ({
   type: 'EDIT_EXPENSE',
   id,
   updates
  });

  export const startEditExpense = (id,updates) => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() =>{
        dispatch(editExpense(id,updates));
      });
    };
  };

  //SET_EXPENSES
  export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
  });

  export const startSetExpenses = () => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
    //fetching expenses from database and parse them to array
     return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
        const expenses = [];
        snapshot.forEach( (childSnapshot) => {
          expenses.push({
           id: childSnapshot.key,
           ...childSnapshot.val()
         });
        });
        dispatch(setExpenses(expenses));
      });
    };
  };

