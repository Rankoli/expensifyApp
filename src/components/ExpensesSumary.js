import React from 'react';
import { connect } from 'react-redux';
import SelectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';



export const ExpensesSumary = (props) => (
    <div>
    {
      props.expenses.length === 1 ? (
           <h1>Viewing {props.expenses.length} expense totaling {numeral(props.expensesTotal/100).format('$0,0.00')} </h1>
      ) : (

           <h1>Viewing {props.expenses.length} expenses totaling {numeral(props.expensesTotal/100).format('$0,0.00')} </h1>
      )
    }

    </div>
    );


const mapStateToProps = (state) => {
    return {
      expenses: SelectExpenses(state.expenses, state.filters),
      expensesTotal:getExpensesTotal(state.expenses)
    };
  };

  export default connect(mapStateToProps)(ExpensesSumary);