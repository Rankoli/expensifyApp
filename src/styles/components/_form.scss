.form {
    display: flex;
    flex-direction: column;
    > * { //select all childrend inside form
        margin-bottom: $m-size;
    }
}

.form__error {
    margin: 0 0 $m-size 0;  //top right bottom left
    font-style: italic;

}