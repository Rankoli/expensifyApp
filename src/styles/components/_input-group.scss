.input-group {
    display: flex;
    margin-bottom: $m-size;
    flex-direction: column;

    @media (min-width: $desktop-breakpoint) { //larger screen
        margin-bottom: $l-size;
        flex-direction: row;

    }
}

.input-group__item {
    margin-bottom: $s-size;

    @media (min-width: $desktop-breakpoint) { //larger screen
        //margin-right: $s-size;
        margin: 0 $s-size 0 0;
    }
}