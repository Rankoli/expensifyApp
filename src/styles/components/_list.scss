.list-header {
    background: $off-white;
    border: 1px solid darken($color: ($off-white), $amount: 7%);
    columns: $grey;
    display: flex;
    justify-content: space-between;
    padding: $s-size $m-size;
}

.list-body {
    margin-bottom: $m-size;
    @media(min-width: $desktop-breakpoint) {
        margin-bottom: $l-size;
    }
}

.list-item {
    border: 1px solid darken($color: $off-white, $amount: 7%);
    border-top: none;
    color: $dark-grey;
    display: flex;
    flex-direction: column;
    padding: $s-size;
    text-decoration: none;
    transition: background .3s ease;
    &:hover {                         //.list-item:hover
        background: $off-white;
    }
    @media (min-width: $desktop-breakpoint) { //larger screen
        align-items: center;
        flex-direction: row;
        justify-content: space-between;
        padding: $m-size;
    }
}

.list-item--message { //no expenses message
    align-items: center;
    justify-content: center;
    color: $grey;
    padding: $m-size;
    &:hover {
        background: none;
    }
}

.list-item__title { //description
    margin: 0;
    word-break: break-all;

}

.list-item__subtitle { //createdAt
    color: $grey;
    font-size: $font-size-small;

}

.list-item__data {  //amount
    //for mobile
    margin: $s-size 0 0 0;
    @media (min-width:$desktop-breakpoint) {
        margin: 0;
        padding-left: $s-size;
    }
}