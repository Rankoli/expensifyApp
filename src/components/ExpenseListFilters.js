import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
import 'react-dates/initialize';
import {DateRangePicker} from "react-dates";

class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this
            .props
            .dispatch(setStartDate(startDate));
        this
            .props
            .dispatch(setEndDate(endDate));
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    };

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            className="text-input"
                            placeholder="Search expenses"
                            onChange={(event) => {
                            this
                                .props
                                .dispatch(setTextFilter(event.target.value))
                        }}/>
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={(event) => {
                            if (event.target.value === 'date') {
                                this
                                    .props
                                    .dispatch(sortByDate());
                            } else {
                                this
                                    .props
                                    .dispatch(sortByAmount());
                            }
                        }}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId="StartDate"
                            endDate={this.props.filters.endDate}
                            endDateId="endDate"
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange=
                            {() => false}
                            showClearDates={true}/>
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {filters: state.filters};
};

export default connect(mapStateToProps)(ExpenseListFilters);