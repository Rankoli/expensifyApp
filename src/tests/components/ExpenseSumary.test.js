import React from 'react'
import {ExpensesSumary} from '../../components/ExpensesSumary';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';


test('should render ExpensesSumary with 1 expense',() => {
    const wrapper = shallow(<ExpensesSumary expenses = {[expenses[0]]} expensesTotal = {195} />);
    expect(wrapper).toMatchSnapshot();
   });

   test('should render ExpenseList with moltply  expensrs',() => {
       const wrapper = shallow(<ExpensesSumary expenses = {[expenses]} expensesTotal = {114195} />);
       expect(wrapper).toMatchSnapshot();
      });