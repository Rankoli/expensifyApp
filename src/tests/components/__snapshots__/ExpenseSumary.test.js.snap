// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with moltply  expensrs 1`] = `
<div>
  <h1>
    Viewing 
    1
     expense totaling 
    $1,141.95
     
  </h1>
</div>
`;

exports[`should render ExpensesSumary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     expense totaling 
    $1.95
     
  </h1>
</div>
`;
